elf_load.c

```diff
@@ -311,13 +333,224 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
 }
 
 #if CONFIG_LIBVFSCORE
+#if CONFIG_LIBPOSIX_MMAP
+/* If vastart + phdr.p_filesz (vastart) < vastart + phdr.p_memsz (vaend),
+ * 0 out that remainder, either through memset or through anonymous mappings
+ */
+static int elf_load_mmap_filesz_memsz_diff(struct elf_prog *elf_prog, int fd,
+					   GElf_Phdr *phdr __maybe_unused,
+					   uintptr_t vastart, uintptr_t vaend)
+{
+	UK_ASSERT(elf_prog && phdr && vastart && vaend);
+
+	uk_pr_debug("%s: Zeroing 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)(vastart),
+		    (uint64_t)(vaend));
+
+	/* From the last byte contained in filesz to the last
+	 * byte contained in memsz, we can either have:
+	 * 1. a page alignment adjustment, where the end
+	 * address is just PAGE_ALIGN_UP(paddr + filesz), which
+	 * means that our initial call to mmap already read in
+	 * a page for us whose remaining bytes we memset without
+	 * generating a page fault, because
+	 * vaend - vastart < PAGE_SIZE
+	 * ...
+	 */
+	memset((void *)vastart, 0, PAGE_ALIGN_UP(vastart) - vastart);
+
+	if (vaend == PAGE_ALIGN_UP(vastart))
+		return 0;
+
+	/*
+	 * ...
+	 * 2. vaend - vastart >= PAGE_SIZE, which can happen if
+	 * this segment contains a NOBITS section, such as .bss.
+	 * If that is the case, then simply anonymously map this
+	 * remaining area so that we don't waste time memsetting
+	 * it (.bss is quite large).
+	 */
+	vastart = PAGE_ALIGN_UP(vastart);
+	vastart = (uintptr_t)mmap((void *)vastart, vaend - vastart,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+				  fd, 0);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the NOBITS part of phdr at "
+			  "offset %lu\n", phdr->p_offset);
+		return (int)vastart;
+	}
+
+	return 0;
+}
+
+/* Use this to mmap first PT_LOAD */
+static int do_elf_load_fdphdr_0(struct elf_prog *elf_prog,
```

Comment:
The alignment of the first segment is currently ensured, but it still seems mapping one segment at a time.
My idea is to allocate address space for all segments at once, so that there won't be address conflicts for segments later.

For example:
```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr           FileSiz            MemSiz         Flags  Align
  LOAD           0x0000000001a00000 0x0000000001a00000 0x0000000001a00000 0x0000000000000231 0x0000000000000231 R E 0x200000
  LOAD           0x0000000001a01000 0x0000000001a01000 0x0000000001a01000 0x0000000000d056f0 0x0000000000d056f0 R   0x200000
  LOAD           0x0000000002706af0 0x0000000002707af0 0x0000000002707af0 0x0000000000083d70 0x00000000000a0810 RW  0x200000
```
I'd like to alloc address space from `min(VirtAddr_i)` to `max(VirtAddr_i + MemSiz_i)`, similar to the stable branch.  That is `0x1a00000 -> (0x2707af0+0xa0810)` in this example. If this allocation is successful, it is guaranteed that each segment has its virtual-address space. Otherwise, if we alloc the segments one by one, it's possible that the first segment is mapped into a "narrow space", and the other segments may conflict with something from kernel. What do you think of this?

Reply:
In most cases, I think the current implementation is enough. But there's still a little possibility that some applications can not be loaded.

Reply:
> I'd like to alloc address space from `min(VirtAddr_i)` to `max(VirtAddr_i + MemSiz_i)`, similar to the stable branch. That is `0x1a00000 -> (0x2707af0+0xa0810)` in this example. If this allocation is successful, it is guaranteed that each segment has its virtual-address space. Otherwise, if we alloc the segments one by one, it's possible that the first segment is mapped into a "narrow space", and the other segments may conflict with something from kernel. What do you think of this?

The point of this PR was to make it so we use demand paging sooner or later for each segment. If we do an initial huge mmap and then read segments into it, how is it different from the `uk_memalign` version.

> and the other segments may conflict with something from kernel

I don't see the possibility of that happening. Do you have an example in mind? :thinking:

Reply:
> I don't see the possibility of that happening. Do you have an example in mind? 🤔

Simply assume the ELF header as follows:
```
Program Headers:
  Type           VirtAddr           MemSiz
  LOAD          0x0000            0x1000
  LOAD          0x1000             0x1000
  LOAD          0x2000            0x1000
  LOAD          0x3000            0x1000
```

Assume that before loading the application, some kernel libraries called `mmap` during initialization, and occupied `0x10000000->0x10001000` and `0x10002000->0x10008000` for some reason. When we load the ELF, we first try to mmap the first segment, and it is possibly mapped to `0x10001000->0x10002000`, and the va_base is `0x10001000`. Then the 3rd and 4th segment will be conflict with `0x10002000->0x10008000`(occupied by something others), using MAP_FIXED to erase them is also unsafe.

Reply:
> But in that time interval, if that happened in the kernel then... the kernel must not be doing something right, don't you agree? All of the required kernel related setup happens before entering ELF loader.

I think it's legal, and the setup indeed happened before entering ELF loader -- then left the virtual address space with a "empty hole", which is just enough for the first segment. It may happen as more and more libraries (has mmap dependencies) are integrated into kernel.

I also think it's fine for now, marking a TODO will be enough.